File: 42/piscine_CPP/d07/ex01/iter.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   iter.cpp                                           :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: skerkour <skerkour@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2015/06/24 17:22:14 by skerkour          #+#    #+#             */
/*   Updated: 2015/06/24 22:14:24 by skerkour         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <cstdlib>
#include <iostream>
#include <string>

template<typename T>
void	iter(T *arr, size_t len, void (*func)(const T&))
{
	for (size_t i = 0; i < len; ++i) {
		func(arr[i]);
	}
}

template< typename T >
void	print(const T& x)
{
	std::cout << x << std::endl;
}

int main() {
	int a[] = {0, 1, 2, 3, 4};
	std::string b[] = {"salut", "les", "gens", "yo"};

	iter(a, 5, print);
	iter(b, 4, print);

	return 0;
}

File: 42/Piscine_OCaml/D01/ex06/iter.ml
(* ************************************************************************** *)
(*                                                                            *)
(*                                                        :::      ::::::::   *)
(*   iter.ml                                            :+:      :+:    :+:   *)
(*                                                    +:+ +:+         +:+     *)
(*   By: skerkour <skerkour@student.42.fr>          +#+  +:+       +#+        *)
(*                                                +#+#+#+#+#+   +#+           *)
(*   Created: 2016/06/14 14:42:13 by skerkour          #+#    #+#             *)
(*   Updated: 2016/06/15 21:48:22 by skerkour         ###   ########.fr       *)
(*                                                                            *)
(* ************************************************************************** *)

let rec iter fn x n =
  if n < 0 then
    -1
  else if n == 0 then
    x
  else
    iter fn (fn x) (n-1)

let () =
  print_string "Test with [(fun x -> x * x) 2 4] : ";
  print_int (iter (fun x -> x * x) 2 4);
  print_char '\n';
  print_string "Test with [(fun x -> x * 2) 2 4] : ";
  print_int (iter (fun x -> x * 2) 2 4);
  print_char '\n';
  print_string "Test with [(fun x -> x * 2) 2 10] : ";
  print_int (iter (fun x -> x * 2) 2 10);
  print_char '\n';
  print_string "Test with [(fun x -> x * 2) 2 14] : ";
  print_int (iter (fun x -> x * 2) 2 14);
  print_char '\n';
